Patterns/DependencyInjection.md

```diff
@@ -0,0 +1,121 @@
+# Dependency Injection
+
+Dependency Injection (DI) forms the basis of most of the applications build within Triple. It removes the burden of knowing how to construct and new-up dependencies and allows classes to simply ask for stuff and get it from the DI container.
+
+## Fundamentals
+
+- [DI in ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/dependency-injection?view=aspnetcore-6.0)
+- [DI in Azure Functions .NET Core 3.1](https://docs.microsoft.com/en-us/azure/azure-functions/functions-dotnet-dependency-injection)
+- [DI in Azure Functions .NET >= 5](https://docs.microsoft.com/en-us/azure/azure-functions/dotnet-isolated-process-guide)
+
+## Best practices concerning DI
+
+### Always depend on interfaces
+
+Try to request interfaces from the container instead of concrete classes. This makes the coupling between two object more loose, and allows for completely swapping the implementation of an interface without affecting any other class. This also makes mocking in unit tests quite easy, as Moq can mock an interface quite easily.
+
+Example:
+
+```c#
+public class Service : IService 
+{
+    public Service(IRepository repository) 
+    {
+
+    }
+}
+```
+
+The `Service` depends on the interface `IRepository` and not the class `Repository`, making it easy to swap it from `SqlRepository` to `InMemoryRepository` or even to `Mock<IRepository>`. The service itself is available in the container under `IService`, so objects like Controllers or other services that depend on it do not depend on the concrete `Service` but the abstract `IService`.
+
+### Choose the shortest lifetime
+
+When configuring the DI container, one can choose three different lifetimes of a service: Transient, Scoped and Singleton. As a best practice, try to pick the shortest lifetime in which the functionality of the service is not affected.
+
+- Transient: This is for services that do not have state and depend on other objects for getting or setting data.
+
+- Scoped: This is for services that do have some state, and that state is scoped to the request (in the case of ASP.NET Core). `DbContext`s are usually scoped. 
+
+- Singleton: This is for services that have some inner state that is not specific to a user, or are costly to construct and don't really change much during the lifetime of the application.
+
+#### Important
+
+- Don't inject everything as a Singleton because that will save allocations. .NET applications have a garbage collector that allocates quickly and web applications often do have hefty garbage collection runs. It's way more painful to downgrade a Singleton service to a Scoped service and find all the nasty bugs that causes than to just create a new object per request or per use.
+- Don't register a service as Scoped because it depends on a service that is Scoped. If it does not have internal state, register it as Transient.
+- Do register a service using its interface.
+
+### Service factories
```

Comment:
Personally, I don't understand this part.

Reply:
@Nuss9 I've reworked it a bit, can you see if it makes more sense?